Tests/AssemblerLib/LocalToGlobalIndexMapMultiComponent.cpp

```diff
@@ -0,0 +1,290 @@
+
+		MGTL::MeshNodeSearcher& searcher_nodes = MGTL::MeshNodeSearcher::getMeshNodeSearcher(*mesh);
+		MGTL::BoundaryElementsSearcher searcher_elements(*mesh, searcher_nodes);
+
+		auto elems = searcher_elements.getBoundaryElements(*ply);
+
+		// deep copy because the searcher destroys the elements.
+		std::transform(elems.cbegin(), elems.cend(),
+					   std::back_inserter(boundary_elements),
+					   std::mem_fn(&MeL::Element::clone));
+
+		std::vector<MeL::Node*> nodes = MeL::getUniqueNodes(boundary_elements);
+
+		mesh_items_boundary.reset(
+		    mesh_items_all_nodes->getIntersectionByNodes(nodes));
+	}
+
+	virtual ~AssemblerLibLocalToGlobalIndexMapMultiDOFTest()
+	{
+		for (auto p : components_boundary)
+			delete p;
+		for (auto p : components)
+			delete p;
+		for (auto p : boundary_elements)
+			delete p;
+	}
+
+	void clear()
+	{
+		for (auto* mss : components) {
+			delete mss;
+		}
+
+		components.clear();
+
+		for (auto* mss : components_boundary) {
+			delete mss;
+		}
+
+		components_boundary.clear();
+	}
+
+	void initComponents(const unsigned num_components, const unsigned selected_component,
+						const AL::ComponentOrder order)
+	{
+		assert(selected_component < num_components);
+
+		clear();
+
+		for (unsigned i=0; i<num_components; ++i)
+		{
+			components.push_back(new MeL::MeshSubsets(
+			    mesh_items_all_nodes.get()));
+		}
+		dof_map.reset(new AL::LocalToGlobalIndexMap(components, order));
+
+		components_boundary.resize(num_components, nullptr);
+		components_boundary[selected_component] =
+		    new MeL::MeshSubsets(mesh_items_boundary.get());
+
+		dof_map_boundary.reset(
+				dof_map->deriveBoundaryConstrainedMap(components_boundary, boundary_elements)
+				);
+	}
+
+	template <AL::ComponentOrder order>
+	void test(const unsigned num_components, const unsigned selected_component,
+	          std::function<std::size_t(std::size_t, std::size_t)> const
+	              compute_global_index);
+
+	std::unique_ptr<const MeshLib::Mesh> mesh;
+	std::unique_ptr<const MeL::MeshSubset> mesh_items_all_nodes;
+    std::vector<MeL::MeshSubsets*> components_boundary;
+    std::vector<MeL::MeshSubsets*> components;
+
+	GeoLib::GEOObjects geo_objs;
+
+	std::unique_ptr<AL::LocalToGlobalIndexMap> dof_map;
+	std::unique_ptr<AL::LocalToGlobalIndexMap> dof_map_boundary;
+
+	std::unique_ptr<MeL::MeshSubset const> mesh_items_boundary;
+	std::vector<MeL::Element*> boundary_elements;
+};
+
+
+struct ComputeGlobalIndexByComponent
+{
+	std::size_t num_nodes;
+
+	std::size_t operator()(std::size_t const node,
+	                       std::size_t const component) const
+	{
+		return node + component * num_nodes;
+	}
+};
+
+struct ComputeGlobalIndexByLocation
+{
+	std::size_t num_components;
+
+	std::size_t operator()(std::size_t const node,
+	                       std::size_t const component) const
+	{
+		return node * num_components + component;
+	}
+};
+
+
+template <enum AL::ComponentOrder ComponentOrder>
```

Comment:
My fault here:
```
F:\jenkins\workspace\OGS-6\Win-PRs\ogs\Tests\AssemblerLib\LocalToGlobalIndexMapMultiComponent.cpp(166): error C3431: 'ComponentOrder' : a scoped enumeration cannot be redeclared as an unscoped enumeration [F:\jenkins\workspace\OGS-6\Win-PRs\build\Tests\testrunner.vcxproj]
```

@chleh could you remove the `enum` in templat argument list please.